hack/generator/pkg/codegen/pipeline_flatten_resources.go

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/pkg/errors"
+)
+
+// flattenResources flattens any resources directly inside other resources
+func flattenResources() PipelineStage {
+	return MakePipelineStage(
+		"flatten-resources",
+		"Flatten nested resource types",
+		func(ctx context.Context, defs astmodel.Types) (astmodel.Types, error) {
+
+			v := astmodel.MakeTypeVisitor()
+			originalVisitResource := v.VisitResourceType
+			v.VisitResourceType = func(this *astmodel.TypeVisitor, it *astmodel.ResourceType, ctx interface{}) (astmodel.Type, error) {
+				// visit inner types:
+				visited, err := originalVisitResource(this, it, ctx)
+				if err != nil {
+					return nil, err
+				}
+
+				newResource, err := flattenResource(defs, visited)
+				if err != nil {
+					return nil, err
+				}
+
+				return newResource, nil
+			}
+
+			results := make(astmodel.Types)
+
+			for _, def := range defs {
+
+				result, err := v.VisitDefinition(def, nil)
+				if err != nil {
+					return nil, errors.Wrapf(err, "error processing type %v", def.Name())
+				}
+
+				results.Add(*result)
+			}
+
+			return results, nil
+		})
+}
+
+func flattenResource(defs astmodel.Types, t astmodel.Type) (astmodel.Type, error) {
+	if resource, ok := t.(*astmodel.ResourceType); ok {
+
+		changed := false
+
+		// resolve spec
+		{
+			specType, err := defs.FullyResolve(resource.SpecType())
+			if err != nil {
+				return nil, err
+			}
+
+			if specResource, ok := specType.(*astmodel.ResourceType); ok {
```

Comment:
Woah, where is this done? Surprised this is a thing?

Reply:
where is what done?

Reply:
the overall nested resource thing?

Reply:
Yeah, the resource within resource

Reply:
so it’s because I now identify `ResourceBase` as a resource so that child resources get identified as resources by inheriting from it (with `allOf`)

With a resource like `Microsoft.Relay` `Namespace(s)` after extracting from the deployment template it will look like:

```
Namespaces: resource(allOf(ResourceBase, object(...)))
```

Once we start reducing the `allOf` the `ResourceBase` gets inlined:

```
Namespaces: resource(allOf(resource(allOf(ARMResourceBase, object(...))), object(...)))
```

`ARMResourceBase` will also be inlined and is also considered to be a resource, so…

```
Namespaces: resource(allOf(resource(allOf(resource(object(...)), object(...))), object(...)))
```

And that turns into:
```
Namespaces: resource(resource(resource(object(...)))
```

Actually now I'm wondering why this doesn't get fully reduced 😁  I shall poke it a bit more!

Reply:
Ok so I tried a change that we modify `WithSpec` and `WithStatus` to automatically unpack any resources we pass to them and extract the appropriate spec/status part, and this clears up a lot of the nesting, but we still need a “flatten” pass to deal with `TypeName`s since that can only be done once we have a full set of the types assembled. 

So e.g. we can end up with:

```
SomeType: resource(ResourceBase)
ResourceBase: resource(...)
```

Which still needs flattening later.

Reply:
Perhaps marking `resourceBase` types as resources isn't the greatest approach and we need an `looksLikeResource` check that we can perform 🤔

Reply:
I think actually I'm not going to change this. it seems to work as-is and I haven't got a better scheme for it at the moment